### PR TITLE
MNE Deep CNTK integration

### PR DESCRIPTION
**Pull request by chdinh**

### CODE REVIEW

File: MNE/deep/deepmodelcreator.h [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    * Default constructor
+    */
+    DeepModelCreator();
+
+    //=========================================================================================================
+    /**
+    * Destructs DeepModelCreator
+    */
+    virtual ~DeepModelCreator();
+
```
Comment (LorenzE) [inline MNE/deep/deepmodelcreator.h]: remove empty line brudaaa

File: MNE/deep/deepmodelcreator.h [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+//    */
+//    static CNTK::FunctionPtr SetupFullyConnectedDNNLayer(CNTK::Variable input, size_t outputDim, const CNTK::DeviceDescriptor &device, const std::function<CNTK::FunctionPtr (const CNTK::FunctionPtr &)> &nonLinearity)
+//    {
+//        return nonLinearity(SetupFullyConnectedLinearLayer(input, outputDim, device));
+//    }
+
+
+
+
+    static CNTK::FunctionPtr FullyConnectedFeedForwardClassifierNet(CNTK::Variable input,
```
Comment (LorenzE) [inline MNE/deep/deepmodelcreator.h]: mind lowercase

File: MNE/deep/deepmodelcreator.h [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                                                                    const CNTK::Parameter& inputTimesParam,
+                                                                    const CNTK::Parameter& inputPlusParam,
+                                                                    const CNTK::Parameter hiddenLayerTimesParam[],
+                                                                    const CNTK::Parameter hiddenLayerPlusParam[],
+                                                                    const CNTK::Parameter& outputTimesParam,
+                                                                    const std::function<CNTK::FunctionPtr(const CNTK::FunctionPtr&)>& nonLinearity);
+
+
+private:
+    static CNTK::FunctionPtr FullyConnectedDNNLayer(CNTK::Variable input,
```
Comment (LorenzE) [inline MNE/deep/deepmodelcreator.h]: mind lowercase

File: MNE/disp/deepmodelviewer/controls.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE MEMBER METHODS
+//=============================================================================================================
+
+Controls::Controls(QWidget* parent)
+: QWidget(parent)
+, m_view(Q_NULLPTR)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/controls.cpp]: m_pView alta

File: MNE/disp/deepmodelviewer/controls.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    QRadioButton *roundCap = new QRadioButton(m_capGroup);
+    m_capGroup->setTitle(tr("Cap Style"));
+    flatCap->setText(tr("Flat"));
+    squareCap->setText(tr("Square"));
+    roundCap->setText(tr("Round"));
+    flatCap->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+    squareCap->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+    roundCap->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+
+    m_joinGroup = new QGroupBox(parent);
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/controls.cpp]: m_pJoinGroup. please change for other member variables as well

File: MNE/disp/deepmodelviewer/controls.h [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE NAMESPACE DISPLIB
+//=============================================================================================================
+
+namespace DISPLIB
+{
+
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/controls.h]: add empty line

File: MNE/disp/deepmodelviewer/controls.h [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    Controls(QWidget *parent = Q_NULLPTR);
+
+    Controls(View* v, QWidget *parent = Q_NULLPTR);
+
+    void setView(View* v);
+
+private:
+    View* m_view;                   /**< The view which this control is connected to */
+
+    QGroupBox *m_capGroup;
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/controls.h]: m_p...

File: MNE/disp/deepmodelviewer/node.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+static double TwoPi = 2.0 * Pi;
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE MEMBER METHODS
+//=============================================================================================================
+
+Node::Node(DeepViewerWidget *graphWidget)
+: graph(graphWidget)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/node.cpp]: m_pGraph

File: MNE/disp/deepmodelviewer/node.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+, m_diameter(20)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/node.cpp]: m_dDiameter

File: MNE/disp/deepmodelviewer/node.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+, m_radius(m_diameter/2)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/node.cpp]: m_fRadius

File: MNE/disp/deepmodelviewer/node.h [hunk 11]
```diff
@@ -32,99 +29,91 @@
...
 
-    //FIFF File management
-    QString                 m_fiffFileName;
+    void mousePressEvent(QGraphicsSceneMouseEvent *event) override;
+    void mouseReleaseEvent(QGraphicsSceneMouseEvent *event) override;
 
-    //QFile m_fiffFile;
+private:
+    float m_diameter;
+    float m_radius; //half diameter -> for drawing speed already calculated beforehand
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/node.h]: m_dRadius, m_fDiameter

File: MNE/disp/deepmodelviewer/node.h [hunk 12]
```diff
@@ -32,99 +29,91 @@
...
+    QList<Edge *> edgeList;
+    QPointF newPos;
+    DeepViewerWidget *graph;
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/node.h]: m_pGraph

File: MNE/disp/deepmodelviewer/view.cpp [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+#endif
+
+
+//*************************************************************************************************************
+
+View::View(const QString &name, QWidget *parent)
+: QWidget(parent)
+, m_use_antialiasing(true)
+#ifndef QT_NO_OPENGL
+, m_use_opengl(false)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/view.cpp]: m_bUseOpenGL

File: MNE/disp/deepmodelviewer/view.cpp [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+    }
+}
+#endif
+
+
+//*************************************************************************************************************
+
+View::View(const QString &name, QWidget *parent)
+: QWidget(parent)
+, m_use_antialiasing(true)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/view.cpp]: m_bUseAntialiasing

File: MNE/disp/deepmodelviewer/view.cpp [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+
+    QGridLayout *topLayout = new QGridLayout;
+    topLayout->addLayout(labelLayout, 0, 0);
+    topLayout->addWidget(graphicsView, 1, 0);
+    topLayout->addLayout(zoomSliderLayout, 1, 1);
+    topLayout->addLayout(rotateSliderLayout, 2, 0);
+    topLayout->addWidget(resetButton, 2, 1);
+    setLayout(topLayout);
+
+    connect(resetButton, SIGNAL(clicked()), this, SLOT(resetView()));
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/view.cpp]: sry but SLOT SIGNAL syntax are not used anymore!! brudda cmmonn!!

File: MNE/disp/deepmodelviewer/view.cpp [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+
+    setupMatrix();
+
+    enableOpenGL(m_use_opengl);
+}
+
+
+//*************************************************************************************************************
+
+QGraphicsView *View::view() const
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/view.cpp]: please rename. this reads like the contsructor. maye getView????

File: MNE/disp/deepmodelviewer/view.cpp [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+                              ? QGraphicsView::RubberBandDrag
+                              : QGraphicsView::ScrollHandDrag);
+    graphicsView->setInteractive(selectModeButton->isChecked());
+}
+
+
+//*************************************************************************************************************
+
+#ifndef QT_NO_OPENGL
+void View::enableOpenGL(bool use_opengl)
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/view.cpp]: please mind this ____ nonsense in the whole file

File: MNE/disp/deepmodelviewer/view.h [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+};
+
+
+//=============================================================================================================
+/**
+* The Deep model View Widget containing the graphics view, as well as view related functions
+*
+* @brief The View Widget containing the graphics view
+*/
+class View : public QWidget
```
Comment (LorenzE) [inline MNE/disp/deepmodelviewer/view.h]: lol ok I am not even going to boher to comment on this ....

File: MNE/disp/lineplot.cpp [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+LinePlot::~LinePlot()
+{
+
+}
+
+
+//*************************************************************************************************************
+
+QChartView *LinePlot::linePlot(const QVector<double> &y, const QString &title)
```
Comment (LorenzE) [inline MNE/disp/lineplot.cpp]: constructor

File: MNE/rtProcessing/rthpis.cpp [hunk 20]
```diff
@@ -330,7 +330,7 @@ Eigen::MatrixXd fminsearch(Eigen::MatrixXd pos,
     while ((func_evals < maxfun) && (itercount < maxiter)) {
 
         for (int i = 0;i < n;i++) {
-            tempX1(i) = abs(fv[0] - fv[i+1]);
+            tempX1(i) = fabs(fv[0] - fv[i+1]);
```
Comment (LorenzE) [inline MNE/rtProcessing/rthpis.cpp]: why u messing with my code

File: applications/mne_analyze/extensions/deepcntk/deepcntk.cpp [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+using namespace CNTK;
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// STATIC FUNCTIONS
+//=============================================================================================================
+
+// Helper function to generate a random data sample
+void generate_random_data_samples(int sample_size, int feature_dim, int num_classes, MatrixXf& X, MatrixXf& Y)
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/deepcntk/deepcntk.cpp]: underscores

File: applications/mne_analyze/extensions/deepcntk/deepcntk.cpp [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE MEMBER METHODS
+//=============================================================================================================
+
+DeepCNTK::DeepCNTK()
+: m_controlPanel(Q_NULLPTR)
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/deepcntk/deepcntk.cpp]: m_p...

File: applications/mne_analyze/extensions/deepcntk/deepcntk.cpp [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+    // Initialize the parameters for the trainer
+    int minibatch_size = 25;
+    int num_samples = 20000;
+
+    MatrixXf features, labels;
+
+    QVector<double> vecLoss, vecError;
+    generate_random_data_samples(num_samples, static_cast<int>(input_dim), static_cast<int>(num_output_classes), features, labels);
+    m_pDeep->trainModel(features, labels, vecLoss, vecError, minibatch_size, device);
+
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/deepcntk/deepcntk.cpp]: remove empty lines

File: applications/mne_analyze/extensions/deepcntk/deepcntk.h [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// FORWARD DECLARATIONS
+//=============================================================================================================
+
+namespace DISPLIB
+{
+class DeepViewerWidget;
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/deepcntk/deepcntk.h]: please mind indentation

File: applications/mne_analyze/extensions/deepcntk/deepcntk.h [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+
+protected:
+
+private:
+
+    // Deep Model
+    QSharedPointer<DEEPLIB::Deep>   m_pDeep;            /**< CNTK Wrapper */
+    CNTK::FunctionPtr               m_pModel;           /**< CNTK Model */
+
+    DISPLIB::Controls*              m_controlPanel;     /**< View Control Panel */
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/deepcntk/deepcntk.h]: m_pControlPanel. please scan below too

File: applications/mne_analyze/extensions/fiffio/fiffio.h [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+namespace FIFFIOEXTENSION
+{
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// USED NAMESPACES
+//=============================================================================================================
+
+using namespace ANSHAREDLIB;
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/fiffio/fiffio.h]: no using namespace in header pls!

File: applications/mne_analyze/extensions/surfer/surfer.cpp [hunk 27]
```diff
@@ -32,117 +29,139 @@
...
+using namespace ANSHAREDLIB;
+
 
 //*************************************************************************************************************
 //=============================================================================================================
 // DEFINE MEMBER METHODS
 //=============================================================================================================
 
+Surfer::Surfer()
+: m_control(NULL)
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/surfer/surfer.cpp]: m_pControl and so on

File: applications/mne_analyze/extensions/surfer/surfer.h [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+namespace SURFEREXTENSION
+{
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// USED NAMESPACES
+//=============================================================================================================
+
+using namespace ANSHAREDLIB;
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/surfer/surfer.h]: no using namespace in header

File: applications/mne_analyze/extensions/surfer/surfer.pro [hunk 29]
```diff
@@ -0,0 +1,107 @@
+#--------------------------------------------------------------------------------------------------------------
+#
+# @file     surfer.pro
+# @author   Christoph Dinh <chdinh@nmr.mgh.harvard.edu>;
+#           Matti Hamalainen <msh@nmr.mgh.harvard.edu>
+# @version  1.0
+# @date     July, 2017
```
Comment (LorenzE) [inline applications/mne_analyze/extensions/surfer/surfer.pro]: July? Wow you work in da future?

File: applications/mne_analyze/libs/anShared/Management/extensionmanager.h [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+/**
+* DECLARE CLASS ExtensionManager
+*
+* @brief The ExtensionManager class provides a dynamic plugin loader. As well as the handling of the loaded extensions.
+*/
+class ANSHAREDSHARED_EXPORT ExtensionManager : public QPluginLoader
+{
+    Q_OBJECT
+
+//    friend class MainWindow;
```
Comment (LorenzE) [inline applications/mne_analyze/libs/anShared/Management/extensionmanager.h]: remove

File: applications/mne_analyze/libs/anShared/Management/extensionmanager.h [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+    //=========================================================================================================
+    /**
+    * Returns vector containing all plugins.
+    *
+    * @return reference to vector containing all plugins.
+    */
+    inline const QVector<IExtension*>& getExtensions();
+
+private:
+    QVector<IExtension*>    m_qVecExtensions;       /**< Vector containing all extensions. */
```
Comment (LorenzE) [inline applications/mne_analyze/libs/anShared/Management/extensionmanager.h]: do you delete this ?

File: applications/mne_analyze/mne_analyze/mainwindow.cpp [hunk 32]
```diff
@@ -0,0 +1,323 @@
...
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE MEMBER METHODS
+//=============================================================================================================
+
+MainWindow::MainWindow(QWidget *parent)
+: QMainWindow(parent)
+, m_pExtensionManager(new ANSHAREDLIB::ExtensionManager(this))
+, m_mdiView(Q_NULLPTR)
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.cpp]: m_pMdiView

File: applications/mne_analyze/mne_analyze/mainwindow.h [hunk 33]
```diff
@@ -0,0 +1,158 @@
...
+namespace MNEANALYZE
+{
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE FORWARD DECLARATIONS
+//=============================================================================================================
+
+class MainWindow : public QMainWindow
```
Comment (LorenzE) [inline applications/mne_analyze/mne_analyze/mainwindow.h]: please docu this class. THX buddy

File: applications/mne_scan/plugins/reference/reference.cpp [hunk 34]
```diff
@@ -63,7 +63,7 @@ Reference::Reference()
 : m_bIsRunning(false)
 , m_pRefInput(NULL)
```
Comment (LorenzE) [inline applications/mne_scan/plugins/reference/reference.cpp]: Q_NULLPTR ?

File: applications/mne_scan/plugins/reference/reference.h [hunk 35]
```diff
@@ -147,7 +147,7 @@ class REFERENCESHARED_EXPORT Reference : public SCSHAREDLIB::IAlgorithm
     QSharedPointer<ReferenceToolbarWidget>              m_pRefToolbarWidget;            /**< flag whether thread is running.*/
     QAction*                                            m_pActionRefToolbarWidget;      /**< flag whether thread is running.*/
 
-    IOBUFFER::CircularMatrixBuffer<double>::SPtr        m_pRefBuffer;                   /**< Holds incoming data.*/
+    QSharedPointer<IOBUFFER::_double_CircularMatrixBuffer>  m_pRefBuffer;                   /**< Holds incoming data.*/
```
Comment (LorenzE) [inline applications/mne_scan/plugins/reference/reference.h]: NAISE Fix

File: MNE/rtProcessing/rthpis.cpp (hunk 20 above)
Comment (chdinh) [inline MNE/rtProcessing/rthpis.cpp]: Because!!

File: MNE/disp/lineplot.cpp (hunk 19 above)
Comment (chdinh) [inline MNE/disp/lineplot.cpp]: should be actual a constructor class needs to be rewritten in a probable manner

**Pull request merged**